src/baseline/expr.rs

```diff
@@ -778,14 +778,34 @@ where
         if let Some(intrinsic) = self.intrinsic(e.id) {
             self.emit_intrinsic_bin(&e.lhs, &e.rhs, dest, intrinsic, Some(e.op));
         } else if e.op == BinOp::Cmp(CmpOp::Is) || e.op == BinOp::Cmp(CmpOp::IsNot) {
-            self.emit_expr(&e.lhs, REG_RESULT.into());
-            let offset = self.reserve_temp_for_node(&e.lhs);
-            self.asm
-                .store_mem(MachineMode::Ptr, Mem::Local(offset), REG_RESULT.into());
+            let mode = self.ty(e.lhs.id()).mode();
+            let offset;
+            if mode.is_float() {
+                let dest_mode = match mode {
+                    MachineMode::Float32 => MachineMode::Int32,
+                    MachineMode::Float64 => MachineMode::Int64,
+                    _ => unreachable!()
+                };
+                self.emit_expr(&e.lhs, FREG_RESULT.into());
+                self.asm.float_as_int(dest_mode, REG_RESULT, mode, FREG_RESULT);
+                offset = self.reserve_temp_for_node(&e.lhs);
+                self.asm
+                    .store_mem(MachineMode::Ptr, Mem::Local(offset), REG_RESULT.into());
 
-            self.emit_expr(&e.rhs, REG_TMP1.into());
-            self.asm
-                .load_mem(MachineMode::Ptr, REG_RESULT.into(), Mem::Local(offset));
+                self.emit_expr(&e.rhs, FREG_RESULT.into()); // result is a float
+                self.asm.float_as_int(dest_mode, REG_TMP1, mode, FREG_RESULT);
+                self.asm
+                    .load_mem(MachineMode::Ptr, REG_RESULT.into(), Mem::Local(offset));
+            } else {
+                self.emit_expr(&e.lhs, REG_RESULT.into());
+                offset = self.reserve_temp_for_node(&e.lhs);
+                self.asm
+                    .store_mem(MachineMode::Ptr, Mem::Local(offset), REG_RESULT.into());
+
+                self.emit_expr(&e.rhs, REG_TMP1.into());
+                self.asm
+                    .load_mem(MachineMode::Ptr, REG_RESULT.into(), Mem::Local(offset));
+            }
 
             self.asm.cmp_reg(MachineMode::Ptr, REG_RESULT, REG_TMP1);
```

Comment:
But IIUC this line isn't correct anymore. `cmp_reg` uses always mode `Ptr` which was correct before since only references were compared. But now `===` also works for int, float or byte.

Reply:
Thanks, I updated the code!

It feels a bit clunky, but I couldn't find an existing function like this anywhere. Let me know if you want me to factor out the mode computation into some function.